contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol

```diff
@@ -0,0 +1,195 @@
+     * @param _recipient The address that receives the token
+     * @param _tokenId The tokenId being claimed
+     * @dev ECDSA signatures are used to verify the permission to claim a NFT
+     */
+    function claimNFT(
+        bytes memory _sig,
+        uint256 _blockExpiry,
+        address _recipient,
+        uint256 _tokenId
+    ) public virtual {
+        bytes32 message = getClaimSigningHash(
+            _blockExpiry,
+            _recipient,
+            _tokenId
+        ).toEthSignedMessageHash();
+        address signVerifier = signVerifierRegistry.get(signVerifierId);
+        require(
+            ECDSA.recover(message, _sig) == signVerifier,
+            "Permission to call this function failed"
+        );
+        require(block.number < _blockExpiry, "Sig expired");
+
+        address from = ownerOf(_tokenId);
+        require(from != address(0));
+
+        claimNonces[_recipient]++;
+
+        _safeTransfer(from, _recipient, _tokenId, "");
+    }
+
+    /**
+     * @notice Returns the current claim nonce of a address
+     * @param _recipient The address whose nonce is being fetched
+     * @return nonce The addresses current nonce
+     * @dev This view exposes the nonce as it will be required for the signature.
+     * By including a nonce in the signature and updating the nonce on every claim,
+     * we prevent replay signature attacks, as signatures can only be used once.
+     */
+    function getClaimNonce(address _recipient)
+        external
+        view
+        virtual
+        returns (uint256)
+    {
+        return claimNonces[_recipient];
+    }
+
+    /**
+     * @notice Returns the hash that we expect was signed in claimNFT.
+     * @param _blockExpiry As of which block the signature is no longer valid
+     * @param _recipient The address who receives the token
+     * @param _tokenId The tokenId being claimed
+     * @return hash A bytes32 hash that is signed by the signVerifier
+     * @dev claimNFT uses this view to get the expected message to have been signed.
+     */
+    function getClaimSigningHash(
+        uint256 _blockExpiry,
+        address _recipient,
+        uint256 _tokenId
+    ) public view virtual returns (bytes32) {
+        return
+            keccak256(
+                abi.encodePacked(
+                    _blockExpiry,
+                    _recipient,
+                    _tokenId,
+                    address(this),
+                    claimNonces[_recipient]
+                )
+            );
+    }
+
+    /**
+     * @notice Updates the sign verifier registry address
+     * @param _signVerifierRegistry The address the new registry
+     * @dev Requires the DEFAULT_ADMIN_ROLE to call
+     */
+    function setSignVerifierRegistry(address _signVerifierRegistry)
+        external
+        virtual
+        onlyAdmin
+    {
+        signVerifierRegistry = ISignVerifierRegistry(_signVerifierRegistry);
```

Comment:
Is it necessary for `claimNonces` to be reset in any way when the registry is updated or is that out-of-scope / not needed?

Reply:
If the registry is updated and the `signVerifier` changes, all existing signatures are invalidated so I don't see the a security risk if we don't reset `claimNonces`.

If the registry is updated and the `signVerifier` stays the same, then I would assume the intended behaviour is to simply migrate to a new registry without impacting the existing sigs!

So I don't think we need to reset `claimNonces`

Reply:
Roger that–SGTM. May be worth clarifying this in a documentation `@dev` note, just for posterity.